### ISSUE TITLE
torch.compile generating invalid convolution

### ISSUE THREAD

**Issue by Skylion007** [pytorch/pytorch#103191]
### 🐛 Describe the bug

I am trying to call torch.compile through mosaicml composer's in the diffusion library. When I disable FSDP in the library, I get this compile error for a rather trivial stride pattern (stride=[1,1]).

### Error logs

```
TorchRuntimeError: Failed running call_function <built-in method conv2d of type
object at 0x7f7319fc60e0>(*(FakeTensor(..., device='cuda:6', size=(32, 3, 256,
256), dtype=torch.float16), Parameter(FakeTensor(..., device='cuda:6',
size=(0,), dtype=torch.float16)), Parameter(FakeTensor(..., device='cuda:6',
size=(0,), dtype=torch.float16)), (1, 1), (1, 1), (1, 1), 1), **{}):
expected stride to be a single integer value or a list of -1 values to match the
convolution dimensions, but got stride=[1, 1]

from user code:
   File "/usr/lib/python3/dist-packages/torch/nn/modules/conv.py", line 463, in
forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/usr/lib/python3/dist-packages/torch/nn/modules/conv.py", line 459, in
_conv_forward
    return F.conv2d(input, weight, bias, self.stride,

Set torch._dynamo.config.verbose=True or TORCHDYNAMO_VERBOSE=1 for more
information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```
It would be great it torch.compile allowed stride to be specified as a list where all the values are the same value (instead of them all just being -1).

### Versions

nightly 

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by Skylion007** [comment]
The issue definitely orignates from the logic in this header not dealing with negative expected_dim:
`aten/src/ATen/native/utils/ParamUtils.h`

**Comment by zou3519** [comment]
@Skylion007 do you have a script that reproduces the problem? I attempted the following and it does work:
 ```
import torch
import torch.nn.functional as F

x = torch.randn(32, 3, 16, 16)
weight = torch.randn(3, 3, 4, 4, requires_grad=True)

@torch.compile(backend='aot_eager')
def f(x, weight):
    return F.conv2d(x, weight, stride=(1, 1))

f(x, weight)
```

**Comment by Skylion007** [comment]
It's a little bit complicated to reproduce, talking to my teammates to see if there is a easier way, but the model arch is from https://github.com/mosaicml/diffusion with the inductor config set to "fallback_random: true". Is there anyway I can provide a better log? The minimal IR generate script didn't seem to output anything and the error doesn't have a better trace.

It eval mode works at least when using FSDP tensors for the optimizer, but when switching to normal tensors this error appears even in eval mode.

**Comment by zou3519** [comment]
Maybe try some environment variables? https://pytorch.org/docs/stable/dynamo/troubleshooting.html
TORCHDYNAMO_REPRO_AFTER="dynamo" and TORCHDYNAMO_REPRO_AFTER="aot" seem like they could help

**Comment by Skylion007** [comment]
@zou3519 Unfortunately, it looks like AOT is the thing that is failing:
```
Traceback (most recent call last):
  File "/diffusion/run.py", line 26, in <module>
    main()
  File "/usr/lib/python3/dist-packages/hydra/main.py", line 94, in decorated_main
    _run_hydra(
  File "/usr/lib/python3/dist-packages/hydra/_internal/utils.py", line 394, in _run_hydra
    _run_app(
  File "/usr/lib/python3/dist-packages/hydra/_internal/utils.py", line 457, in _run_app
    run_and_report(
  File "/usr/lib/python3/dist-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/usr/lib/python3/dist-packages/hydra/_internal/utils.py", line 220, in run_and_report
    return func()
  File "/usr/lib/python3/dist-packages/hydra/_internal/utils.py", line 458, in <lambda>
    lambda: hydra.run(
  File "/usr/lib/python3/dist-packages/hydra/_internal/hydra.py", line 132, in run
    _ = ret.return_value
  File "/usr/lib/python3/dist-packages/hydra/core/utils.py", line 260, in return_value
    raise self._return_value
  File "/usr/lib/python3/dist-packages/hydra/core/utils.py", line 186, in run_job
    ret.return_value = task_function(task_cfg)
  File "/diffusion/run.py", line 22, in main
    return train(config)
  File "/diffusion/diffusion/train.py", line 132, in train
    return eval_and_then_train()
  File "/diffusion/diffusion/train.py", line 129, in eval_and_then_train
    trainer.eval()
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2656, in eval
    self._eval_loop(
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2780, in _eval_loop
    self.state.outputs = self._original_model.eval_forward(self.state.batch)
  File "/diffusion/diffusion/models/stable_diffusion.py", line 195, in eval_forward
    unet_out, noise, timesteps = self.forward(batch)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/eval_frame.py", line 292, in _fn
    return fn(*args, **kwargs)
  File "/diffusion/diffusion/models/stable_diffusion.py", line 167, in forward
    latents = self.vae.encode(inputs.half())['latent_dist'].sample().data
  File "/usr/lib/python3/dist-packages/diffusers/utils/accelerate_utils.py", line 44, in wrapper
    if hasattr(self, "_hf_hook") and hasattr(self._hf_hook, "pre_forward"):
  File "/usr/lib/python3/dist-packages/torch/_dynamo/eval_frame.py", line 445, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/convert_frame.py", line 527, in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/convert_frame.py", line 415, in transform
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2024, in run
    super().run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1139, in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars.items)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1099, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/nn_module.py", line 717, in call_function
    ).call_function(tx, [self] + list(args), kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1139, in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars.items)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 333, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1099, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/nn_module.py", line 717, in call_function
    ).call_function(tx, [self] + list(args), kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1139, in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars.items)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 333, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1099, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 333, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 287, in call_function
    return super().call_function(tx, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/functions.py", line 120, in call_function
    return tx.inline_user_function_return(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 594, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2129, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 2226, in inline_call_
    tracer.run()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 707, in run
    and self.step()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 667, in step
    getattr(self, inst.opname)(inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 389, in wrapper
    return inner_fn(self, inst)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 1099, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/usr/lib/python3/dist-packages/torch/_dynamo/symbolic_convert.py", line 558, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/torch.py", line 600, in call_function
    tensor_variable = wrap_fx_proxy(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/builder.py", line 1075, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/variables/builder.py", line 1110, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 1298, in get_fake_value
    raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 1266, in get_fake_value
    return wrap_fake_exception(
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 860, in wrap_fake_exception
    return fn()
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 1267, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 1332, in run_node
    raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
  File "/usr/lib/python3/dist-packages/torch/_dynamo/utils.py", line 1319, in run_node
    return node.target(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/_subclasses/fake_tensor.py", line 1161, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/usr/lib/python3/dist-packages/torch/_subclasses/fake_tensor.py", line 1381, in dispatch
    op_impl_out = op_impl(self, func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/_subclasses/fake_tensor.py", line 612, in conv
    conv_backend = torch._C._select_conv_backend(**kwargs)
``` 
This is the full stack trace if this is helpfu before the exception if this is helpful. I added both those env variables and nothing gets outputed (likely because AOT fails to complete)

**Comment by Skylion007** [comment]
@zou3519 I introduced an error by disable static_weight_shapes, and that properly saved a minified_repro (that errored due to too many free variables due to the lack static_weight_shapes) so it looks like the fail is occurring before it even gets to the minified repro stage. Is there any options I can pass to generate you a proper repro?

**Comment by Skylion007** [comment]
This error is fixed, but there a still of tone of other errors on this model including some other bugs I have filed. Sorry for the delay missed the initial notification, I finally cycled back to this. 

Try running this yaml is a repro for running it on synthetic dataset:
```
run_name: synthetic-sd2-base-256
cluster: r8z3
gpu_num: 8
image: mosaicml/pytorch_vision:2.0.1_cu118-python3.10-ubuntu20.04
integrations:
  - integration_type: "git_repo"
    git_repo: Landanjs/diffusion
    git_branch: landan/synthetic_dataset
    pip_install: .[all]
scheduling:
    resumable: true # Makes job preemptible.
    priority: low # Allows higher priority jobs to preempt you
command: |
  cd diffusion
  HYDRA_FULL_ERROR=1 composer run.py --config-path /mnt/config --config-name parameters
parameters:
  project: landan-test
  batch_size: 2048
  seed: 42
  scale_schedule_ratio: 1.0
  name: synthetic-SD2-base-256
  eval_first: false
  algorithms:
    low_precision_groupnorm:
      attribute: unet
      precision: amp_fp16
    low_precision_layernorm:
      attribute: unet
      precision: amp_fp16
  model:
    _target_: diffusion.models.models.stable_diffusion_2
    pretrained: false
    precomputed_latents: false
    encode_latents_in_fp16: true
    fsdp: true
    val_metrics:
      - _target_: torchmetrics.MeanSquaredError
    val_guidance_scales: []
    loss_bins: []
  dataset:
    train_batch_size: ${batch_size}
    eval_batch_size: 256 # Should be 8 per device -> can we make this automatic?
    train_dataset:
      _target_: diffusion.datasets.laion.synthetic_laion.build_synthetic_laion_dataloader
      batch_size: ${batch_size}
      image_size: 256
      prefetch_factor: 2
      num_workers: 8
      persistent_workers: true
      pin_memory: true
    eval_dataset:
      _target_: diffusion.datasets.laion.synthetic_laion.build_synthetic_laion_dataloader
      batch_size: 8
      image_size: 256
      prefetch_factor: 2
      num_workers: 8
      persistent_workers: True
      pin_memory: True
  optimizer:
    _target_: torch.optim.AdamW
    lr: 1.0e-4
    weight_decay: 0.01
  scheduler:
    _target_: composer.optim.MultiStepWithWarmupScheduler
    t_warmup: 10000ba
    milestones:
      - 200ep
  logger:
    wandb:
      _target_: composer.loggers.wandb_logger.WandBLogger
      name: ${name}
      project: ${project}
      group: ${name}
  callbacks:
    speed_monitor:
      _target_: composer.callbacks.speed_monitor.SpeedMonitor
      window_size: 10
    lr_monitor:
      _target_: composer.callbacks.lr_monitor.LRMonitor
    memory_monitor:
      _target_: composer.callbacks.memory_monitor.MemoryMonitor
    runtime_estimator:
      _target_: composer.callbacks.runtime_estimator.RuntimeEstimator
    optimizer_monitor:
      _target_: composer.callbacks.OptimizerMonitor
  trainer:
    _target_: composer.Trainer
    device: gpu
    max_duration: 550000ba
    eval_interval: 2000ba
    device_train_microbatch_size: 32
    run_name: ${name}
    seed: ${seed}
    scale_schedule_ratio: ${scale_schedule_ratio}
    save_folder: oci://mosaicml-internal-checkpoints/landan/diffusion/test_synthetic  # Insert path to save folder or bucket
    save_interval: 1000ba
    save_overwrite: false
    autoresume: true
    fsdp_config:
      sharding_strategy: "SHARD_GRAD_OP"
```

**Comment by ezyang** [comment]
Is there any way to manually run the effective script with a different PyTorch version (I see there's a docker image hardcoded here). In particular, I want to run it with PT2 nightly.

**Comment by Skylion007** [comment]
This is fixed on 2.1.0
